### ISSUE TITLE
Frontier build times: Scope to improve? (Low priority)

### ISSUE THREAD

**Issue by sarats** [E3SM-Project/E3SM#5449]
Looking at build times of relevant configurations on Frontier, we need to decide if some of these files should be added to a Depends list to lower optimization etc. The build times are not terrible, however reducing unnecessary overhead can improve developer workflow.

MMF: WCYCL1950-MMF1 (851 sec)
https://pace.ornl.gov/buildtime/128983

MMF: F2010-MMF1 (587 sec)
https://pace.ornl.gov/buildtime/128969

SCREAM: F2010-SCREAMv1 (461 sec)
https://pace.ornl.gov/buildtime/128973

**Comment by philipwjones** [comment]
I've run into this with standalone ocean builds on Crusher with cray too. The MPAS top-level ocn_core_interface routine contains a massive include file (> 100k lines) that defines all structs and variables. The mpas streams manager file is also somewhat large with a lot of interfaces. Neither are performance critical and could be compiled at lower optimization but don't know whether it would help - worth a try.

**Comment by sarats** [comment]
@peterdschwartz / @bishtgautam Can you take a look at ELM files and comment if their opt level can be lowered?
```
elm/src/data_types/ColumnDataType.F90.o
elm/src/data_types/ColumnDataType.F90.o
```

**Comment by oksanaguba** [comment]
@ambrad pointed to me that on chrysalis, the 2nd file's opt is already lowered due to the build times.

**Comment by sarats** [comment]
Yes, I suspect a bunch of these files have lowered optimization due to build times on many platforms. I wonder if we should add these kind of data structure style source files to a global NOOPT list to prevent this issue from recurring.

**Comment by sarats** [comment]
@jgfouca Is there a mechanism to have a global Depends file where we are absolutely certain some files need to be added to a NOOPT list everywhere?

**Comment by jgfouca** [comment]
@sarats , yes. Depends files are for machine/compiler specific stuff. If a component wants to make a global change, just do it in the cmake files directly:
https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/1011942171/E3SM+Build+System+and+CMake#E3SMBuildSystemandCMake-Permanentlyaddcompileflagsforaspecificfile

**Comment by ndkeen** [comment]
There is one special case that might be of interest here.

```
In  cime/CIME/Tools/Makefile:

ifeq ($(COMP_NAME),cam)
  # These RRTMG files take an extraordinarily long time to compile with optimization.                                                                                                                      
  # Until mods are made to read the data from files, just remove optimization from                                                                                                                         
  # their compilation.                                                                                                                                                                                     
rrtmg_lw_k_g.o: rrtmg_lw_k_g.f90
        $(FC) -c $(FPPFLAGS) $(INCLDIR) $(INCS) $(FREEFLAGS) $(FFLAGS) $(FFLAGS_NOOPT) $<

rrtmg_sw_k_g.o: rrtmg_sw_k_g.f90
        $(FC) -c $(FPPFLAGS) $(INCLDIR) $(INCS) $(FREEFLAGS) $(FFLAGS) $(FFLAGS_NOOPT) $<

```

**Comment by jgfouca** [comment]
@ndkeen , we don't use the Tools/Makefile for our component builds anymore.

**Comment by ndkeen** [comment]
@jgfouca Those 2 files show reduced optimization for my builds

**Comment by sarats** [comment]
I think we need to build a catalog of files that should be added to the global noop list and add it to CMakeLists.txt (?)

```
if (COMP_NAME STREQUAL "name-of-component")
  list(APPEND NOOPT_FILES "file1" "file2" ...)
endif()
```

**Comment by jgfouca** [comment]
@ndkeen , yes that's because I ported that logic to build_model.cmake (looks you discovered that also). 

@sarats, there's a cmake list NOOPT_FILES that you can use in build_model.cmake.

**Comment by sarats** [comment]
I presume we should append rather than set for subsequent invocations? If so, we should initialize to empty?
```
set(NOOPT_FILES "eam/src/physics/rrtmg/ext/rrtmg_lw/rrtmg_lw_k_g.f90;eam/src/physics/rrtmg/ext/rrtmg_sw/rrtmg_sw_k_g.f90")
```

**Comment by sarats** [comment]
@rljacob Building a list of files that should be added to a NOOPT list from all components sounds like a IG task, agree?

**Comment by rljacob** [comment]
Yes since build things are in IG.

**Comment by sarats** [comment]
FYI. 

Without optimization (Cray compilers),
```
micro_mg_cam.F90 - 3 seconds (vs 191 s earlier)
elm/src/data_types/VegetationDataType.F90 - 8 s (~140 s earlier)
/elm/src/data_types/ColumnDataType.F90 - 8 s (~135 s earlier)
eam/src/physics/cam/micro_mg1_5.F90 - 2 s (45 s earlier)
```

Adding to a Depends file for now while we build the global list.
